### ISSUE TITLE
Find_file() should only create the file if it is selected in telescope

### ISSUE THREAD

**Issue by lambtho12** [renerocksai/telekasten.nvim#20]
In my document, let's say I have [[superinteresting_stuff]] and use the `find_file()` command while this file does not exist in my computer.

As the file does not exist, find_file() will create it first and suggest it in telescope. Now if in telescope I see an other file which already exist and is a better fit for my need, I will likely open this one and never use the newly created `superinteresting_stuff.md`, that will remain blank. This situation also happens if I have a typo in the link I try to open. E.g. [[my fil]] instead of [[my file]]; that would create a useless file as I would pick the proper one in telescope as well. This also happens if I forget that the file I try to reach is in a subdir: `[[thisfile]]` instead of `[[subdir/thisfile]]` would create `thisfile` in the current directory even though I will select the proper file in the telekasten prompt.

In that case, I think telekasten should detect that the file was not opened and remove it when an other file is selected (or when the telescope prompt is closed).

**Comment by renerocksai** [comment]
Good point!

The rationale for creating the note before calling Telescope is: to see the note with the template expanded, as a preview of what you get when you open it. 

While this can be deactivated via a config option, completely de-activating is probably not what you want.

The way I understand it, the best option would be to create the note, have it being pre-viewed in Telescope, but if it is not opened, it should be removed again. This spares you from having to delete notes that you didn't intend to have created.

Challenge accepted :smile:

**Comment by lambtho12** [comment]
> The rationale for creating the note before calling Telescope is: to see the note with the template expanded, as a preview of what you get when you open it.

Yes this is what I understood as well.

> The way I understand it, the best option would be to create the note, have it being pre-viewed in Telescope, but if it is not opened, it should be removed again. This spares you from having to delete notes that you didn't intend to have created.

Yes, this is the idea that I had as well. An other idea would be to assign the file to a new buffer and not actually save a file (let the user save the buffer after modification). That way no file deletion will be required from telekasten. However, I do not know if telescope can display existing files and unsaved buffers in the same search... So this might be actually more complicated and bring its own set of issues.

**Comment by renerocksai** [comment]
> However, I do not know if telescope can display existing files and unsaved buffers in the same search... So this might be actually more complicated and bring its own set of issues.

I fear, the above will be complicated. Buffers and files in the same search.

The more I think of it, maybe we can find a good middle way:

Taking into account the following:
> This also happens if I forget that the file I try to reach is in a subdir: `[[thisfile]]` instead of `[[subdir/thisfile]]` would create `thisfile` in the current directory even though I will select the proper file in the telekasten prompt.

... I arrive at the conclusion that the way telekasten currently checks for existing notes is OK for a scratching-my-own-itch situation where I don't use sub-directories. But it is not OK to expect a file in the root dir, not finding it there, creating it there, and then have Telescope find it in a subdir and remove the unnecessarily created file in the root dir.

Ideally, telekasten should be aware of all subdirs and check for files (no harm if they contain some relative path info) in all of them. Specifically, searching for the thing between the brackets in all subdirs: `[[b/c]]` should be found in `[[a/b/c]]` - only looking for `c` in all dirs might lead to a suboptimal search prompt suggestion pointing to the wrong dir if `c` exists in many dirs. But I digress.

The main point I am trying to make is: telekasten should check all subdirs for files linked to. But how would it know? Directories could be created at any time. So we need to check for subdirs in every follow-link situation. I must check if plenary's `scan_dir()` can handle that. A `find -type d .` command would be very quick as it only needs to traverse the structure of the filesystem, following directory entries and yielding at leaf nodes. If `plenary.scan.scan_dir()` always scans for and returns all files - from which to infer the directory structure, this would be a less efficient fall-back if the `find` command is unavailable.

So I suggest the following to get it more right:

- telekasten checks all subdirs when searching for `filespec`  (coming from `[[` `filespec` `]]`)
  - `filespec` can be a relative path
  - list of subdirs will be obtained via `find` command, with a fallback to plenary's `scan_dir()` unless the latter can scan for dirs only
- if a note is not found, it will be created according to `filespec`
  - interesting question: should non-existing subdirs be created? Option? (if `subdir/note` doesn't exist and `subdir` doesn't exist either; this should be handled gracefully).
- telescope is activated, pointing to the found or created note
- if a just-created note is not picked by telescope, it will be removed

So, in essence, it boils down to removing unnecessarily created notes, combined with avoiding the creation of unnecessary notes as much as possible in the first place.

This will probably take a bit of time to get right. I do have a day job, too. :smile: 

Objections and suggestions are of course welcome. :smile:

**Comment by renerocksai** [comment]
It's done, see the above commit:
- `follow_link()` now searches in all subdirs for what's between the `[[brackets]]`. So it also works for
  `[[subdir/note]]`.
- notes implicitly created by following a link to a non-existing note or current dailies and weeklies will get erased
  if a telescope picker is canceled by <kbd>CTRL</kbd>+<kbd>C</kbd> or <kbd>ESC</kbd>. 

  Unfortunately, I did not find a way to find out whether the picker was canceled. It seems this concept is foreign to
  telescope. While I can determine and be called back when the finder is closed, it is impossible to tell why it was
  closed at that time. So no way to find out if it was closed because a file was picked or the search was canceled.

  For that reason, I mapped the above key combinations to the auto-removal function. If you quit the search by entering
  normal mode and issuing `:q`, auto-removal will be bypassed.

  I just realized: it is very hard to get into normal mode when I mapped the <kbd>ESC</kbd> key in insert mode to the
  close action. This was a bad choice because that deprives one of using normal mode movements and stuff. So I am
  going to remove that binding now...

  :smile:
